### ISSUE TITLE
{horus_ex,module_not_found,#{module => erlang}}

### ISSUE THREAD

**Issue by Zabrane** [rabbitmq/khepri#222]
### What does not work?

Switching from old `khepri` to the latest stable one (`0.7.0`), we're getting this error in the release's logs:
```bash
2023-08-10T11:14:53.997963+00:00 error: Ranch listener http, connection process <0.1643.0>, stream 1 had its request process <0.1645.0> exit with reason {horus_ex,module_not_found,#{module => erlang}} and stacktrace [{horus,get_object_code_from_code_server,1,[{file,"/home/zab/dev/ics/_build/default/lib/horus/src/horus.erl"},{line,1984}]},{horus,disassemble_module1,2,[{file,"/home/zab/dev/ics/_build/default/lib/horus/src/horus.erl"},{line,1927}]},{horus,disassemble_module,2,[{file,"/home/zab/dev/ics/_build/default/lib/horus/src/horus.erl"},{line,1879}]},{horus,lookfunction,4,[{file,"/home/zab/dev/ics/_build/default/lib/horus/src/horus.erl"},{line,1778}]},{horus,pass1_process_function,4,[{file,"/home/zab/dev/ics/_build/default/lib/horus/src/horus.erl"},{line,1155}]},{maps,fold_1,4,[{file,"maps.erl"},{line,416}]},{horus,extract_module_info_functions,1,[{file,"/home/zab/dev/ics/_build/default/lib/horus/src/horus.erl"},{line,745}]},{horus,to_standalone_fun3,2,[{file,"/home/zab/dev/ics/_build/default/lib/horus/src/horus.erl"},{line,497}]}]
[...]
```

### Expected behavior

Our service was perfectly working with old `khepri`.
Moreover, this error doesn't happen locally (DEV mode). It only triggers when deploying our service as a release.

### How to reproduce

1. Erlang `26.0.2`
2. rebar3
```
rebar 2.6.4 22 20210518_220133 git 2.5.1-420-g33b1898
```
3. When connecting to the release node, this [error is raised](https://github.com/rabbitmq/horus/blob/main/src/horus.erl#L1986) IMHO:
```
$ /opt/ics/bin/ics remote_console
Erlang/OTP 26 [erts-14.0.2] [source] [64-bit] [smp:6:6] [ds:6:6:10] [async-threads:1] [jit:ns]
Eshell V14.0.2 (press Ctrl+G to abort, type help(). for help)

> code:get_object_code(erlang).
error

%% but this works as you can see
> erlang:module_info().
[{module,erlang},
 {exports,[{binary_to_atom,1},
           {binary_to_existing_atom,1},
           {binary_to_integer,1},
           {binary_to_integer,2},
           {check_process_code,2},
           {check_process_code,3},
           {alias,0},
           {garbage_collect,1},
           {garbage_collect,2},
           {garbage_collect_message_area,0},
           {is_alive,0},
           {list_to_integer,2},
           {process_display,2},
           {process_flag,3},
           {setnode,3},
           {suspend_process,2},
           {suspend_process,1},
           {trace_pattern,2},
           {spawn,2},
           {spawn_link,2},
           {spawn_monitor,2},
           {spawn_monitor,3},
           {spawn_opt,2},
           {spawn_opt,...},
           {...}|...]},
 {attributes,[]},
 {compile,[]},
 {md5,<<62,137,211,78,85,65,250,44,245,216,197,139,35,74,
        142,40>>}]
        
> User switch command (type h for help)
 --> q        
```
        
_No response_

**Comment by dumbbell** [comment]
Hi!

Do you have a sample release you could share for me to look at?

Indeed, Khepri/Horus should not try to extract the code from the `erlang` module in the first place.

**Comment by Zabrane** [comment]
@dumbbell unfortunately, I've no right (I'm just a developer) to share this release (or source code). But I'm willing to execute any command/script to help you debug this issue.

**Comment by Zabrane** [comment]
@dumbbell would it be possible to exclude the erlang module to be checked from [this call](https://github.com/rabbitmq/horus/blob/main/src/horus.erl#L1986)?

Something along these lines:
```erlang
get_object_code_from_code_server(erlang) ->
%% ignore erlang module
;
get_object_code_from_code_server(Module) ->
    case code:get_object_code(Module) of
        {Module, Beam, Filename} ->
            {Module, Beam, Filename, code_server};
        error ->
            ?horus_misuse(
               module_not_found,
               #{module => Module})
    end.
```

**Comment by dumbbell** [comment]
We already do something like that earlier in the code path. `erlang` is not the only module we want to skip. See `horus_utils:should_process_module/1` and `horus_utils:init_list_of_modules_to_skip/0`.

After you get the error, could you please tell me what the return value of the following call is? Or does it throw an exception?
```erlang
persistent_term:get({horus, skipped_modules_in_code_collection})
```

**Comment by dumbbell** [comment]
Can you wrap the call into `rp(...)` like this if you run it from the shell:
```erlang
rp(persistent_term:get({horus, skipped_modules_in_code_collection}))
```
Otherwise the shell truncates the output.

**Comment by dumbbell** [comment]
Thank you! `erlang` is missing from that list, weird...

Could you please share the output of:
```erlang
[begin
     _ = application:load(App),
     application:get_key(App, modules)
 end || App <- [erts, kernel, stdlib]].     
```

**Comment by Zabrane** [comment]
> Thank you! `erlang` is missing from that list, weird...
> 
> Could you please share the output of:
> 
> ```erlang
> [begin
>      _ = application:load(App),
>      application:get_key(App, modules)
>  end || App <- [erts, kernel, stdlib]].
> ```

```erlang
[undefined,
 {ok,[application,application_controller,application_master,
      application_starter,auth,code,code_server,dist_util,
      erl_boot_server,erl_compile_server,erl_distribution,
      erl_erts_errors,erl_reply,erl_kernel_errors,
      erl_signal_handler,erpc,error_handler,error_logger,file,
      file_server,file_io_server,global,global_group,
      global_search,group|...]},
 {ok,[argparse,array,base64,beam_lib,binary,c,calendar,dets,
      dets_server,dets_sup,dets_utils,dets_v9,dict,digraph,
      digraph_utils,edlin,edlin_context,edlin_expand,
      edlin_type_suggestion,epp,eval_bits,erl_abstract_code,
      erl_anno,erl_bits|...]}]
```

**Comment by dumbbell** [comment]
So your release doesn't have the `erts` application. I thought it was implicit to include it in any release as it is the actual runtime. Could you please list files in your release, after redacting any private filenames?

**Comment by Zabrane** [comment]
Here is the list of our apps bundled in rebar.config
```erlang
%% Release
{relx, [{release, {ics, "latest"},
         [ kernel
         , stdlib
         , sasl
         , crypto
         , asn1
         , inets
         , compiler
         , syntax_tools
         , runtime_tools
         , tools
         %% SERVICES
         , ics
         , jiffy
         , gen_smtp
         , bbmustache
         , sh
         , uuid
         , khepri
         , cowboy
        ]}
```

**Comment by dumbbell** [comment]
And what about the actual list of files in the release archive?

**Comment by Zabrane** [comment]
Adding `erts` app to the list above seems to have fixed the issue.
Now, we don't have this error anymore with `horus`.
```bash
$ ls -1 /opt/ics/lib/erts-14.0.2/ebin/
atomics.beam
counters.beam
erlang.beam <--------- HERE IS THE MISSING GUY
erl_init.beam
erl_prim_loader.beam
erl_tracer.beam
erts.app
erts_code_purger.beam
erts_dirty_process_signal_handler.beam
erts_internal.beam
erts_literal_area_collector.beam
init.beam
persistent_term.beam
prim_buffer.beam
prim_eval.beam
prim_file.beam
prim_inet.beam
prim_net.beam
prim_socket.beam
prim_zip.beam
socket_registry.beam
zlib.beam
```

**Comment by Zabrane** [comment]
@dumbbell Could it be possible to make the `erts` app as a `khepri` dependency to avoid this issue?

**Comment by dumbbell** [comment]
Can I please get the entire list of files from your release, minus what is private, before you change that list?

I need to understand what the release looked like if `erts` isn't explicitly listed. I don't see why it would not be always included as nothing will work without it.

**Comment by Zabrane** [comment]
Sure thing. Here is a diff before and after adding the `erts` app explicitly to the release.

```diff
--- /tmp/a      2023-08-10 13:07:11.624935221 +0000
+++ /tmp/b      2023-08-10 13:07:08.089055463 +0000
@@ -9,6 +9,7 @@
 cowlib-2.11.0
 crypto-5.2
+erts-14.0.2
 gen_batch_server-0.8.8
 gen_smtp-1.3.0-d
 ```

**Comment by dumbbell** [comment]
Do you include the runtime (native executable like `beam` and so on) in the release?

**Comment by Zabrane** [comment]
Yes we do.

**Comment by dumbbell** [comment]
Ok, thank you.

I will probably add `erts` as a dependency to Horus to work around this problem. But it really feels like there is a problem in the way the release is created (probably not something on your side, but rather in relx or reltool). Not including `erts` by default (and always) does not make much sense to me. I will try to reproduce and study that further.

**Comment by Zabrane** [comment]
Thanks a lot for your support @dumbbell 
We are planning to use `khepri` in all our products and staying away from `Mnesia`

**Comment by dumbbell** [comment]
Thank you! That's awesome to know :-)

**Comment by dumbbell** [comment]
Horus 0.2.4 and the just released Khepri 0.8.0 include a fix for this problem.

**Comment by Zabrane** [comment]
@dumbbell We gonna give `Khepri 0.8.0`a try and report back any issue.
